File: Server/app/services/portfolio_service.py
from __future__ import annotations

import secrets
from datetime import datetime, timezone

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession


class PortfolioService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def list_transactions(self, user_id: str) -> list[dict]:
        result = await self.db.execute(
            text(
                """
                SELECT id, symbol, side, quantity::text, price::text, traded_at::text, metadata
                FROM portfolio_transactions
                WHERE user_id = :user_id
                ORDER BY traded_at DESC, created_at DESC
                """
            ),
            {"user_id": user_id},
        )
        return [
            {
                "id": r[0],
                "symbol": r[1],
                "side": r[2],
                "quantity": float(r[3]),
                "price": float(r[4]),
                "traded_at": r[5],
                "metadata": r[6] if r[6] else {},
            }
            for r in result.all()
        ]

    async def create_transaction(
        self, user_id: str, input_data: dict
    ) -> str:
        tid = secrets.token_hex(16)
        traded_at = input_data.get("traded_at") or datetime.now(timezone.utc).isoformat()
        await self.db.execute(
            text(
                """
                INSERT INTO portfolio_transactions (id, user_id, symbol, side, quantity, price, traded_at)
                VALUES (:id, :uid, :symbol, :side, :qty, :price, :traded_at)
                """
            ),
            {
                "id": tid,
                "uid": user_id,
                "symbol": input_data["symbol"],
                "side": input_data["side"],
                "qty": input_data["quantity"],
                "price": input_data["price"],
                "traded_at": traded_at,
            },
        )
        return tid

    async def update_transaction(self, user_id: str, txn_id: str, input_data: dict) -> None:
        await self.db.execute(
            text(
                """
                UPDATE portfolio_transactions
                SET quantity = :qty, price = :price, traded_at = :traded_at
                WHERE id = :id AND user_id = :uid
                """
            ),
            {
                "qty": input_data["quantity"],
                "price": input_data["price"],
                "traded_at": input_data["traded_at"],
                "id": txn_id,
                "uid": user_id,
            },
        )

    async def delete_transaction(self, user_id: str, txn_id: str) -> None:
        await self.db.execute(
            text(
                "DELETE FROM portfolio_transactions WHERE id = :id AND user_id = :uid"
            ),
            {"id": txn_id, "uid": user_id},
        )

    async def get_summary(self, user_id: str, latest_prices: dict[str, float]) -> dict:
        transactions = await self.list_transactions(user_id)
        holdings_map = self._build_holdings_map(transactions, latest_prices)
        holdings = [h for h in holdings_map.values() if h["quantity"] > 0]

        invested = sum(h["avg_price"] * h["quantity"] for h in holdings)
        current = sum(h["current_value"] for h in holdings)
        realized = sum(h["realized_pnl"] for h in holdings_map.values())
        unrealized = sum(h["unrealized_pnl"] for h in holdings)

        return {
            "cards": [
                {"id": "invested", "label": "Total Invested", "value": invested, "change": None},
                {
                    "id": "current",
                    "label": "Current Value",
                    "value": current,
                    "change": (current / invested - 1) * 100 if invested else 0,
                },
                {
                    "id": "total_pl",
                    "label": "Total P&L",
                    "value": realized + unrealized,
                    "change": ((realized + unrealized) / invested) * 100 if invested else 0,
                },
                {
                    "id": "unrealized",
                    "label": "Unrealized",
                    "value": unrealized,
                    "change": (unrealized / invested) * 100 if invested else 0,
                },
                {"id": "realized", "label": "Realized", "value": realized, "change": None},
            ],
            "totals": {
                "invested": invested,
                "current": current,
                "realized": realized,
                "unrealized": unrealized,
                "total_pnl": realized + unrealized,
            },
        }

    async def get_holdings(self, user_id: str, latest_prices: dict[str, float]) -> list[dict]:
        transactions = await self.list_transactions(user_id)
        holdings_map = self._build_holdings_map(transactions, latest_prices)
        rows = [h for h in holdings_map.values() if h["quantity"] > 0]

        result = []
        for h in rows:
            result.append({
                "symbol": h["symbol"],
                "quantity": h["quantity"],
                "avg_price": h["avg_price"],
                "current_price": h["current_price"],
                "current_value": h["current_value"],
                "pnl": h["unrealized_pnl"],
                "pnl_pct": (h["current_price"] - h["avg_price"]) / h["avg_price"] * 100 if h["avg_price"] else 0,
                "realized_pnl": h["realized_pnl"],
                "allocation_pct": 0,
            })

        total_current = sum(r["current_value"] for r in result)
        for r in result:
            r["allocation_pct"] = (r["current_value"] / total_current * 100) if total_current else 0

        return result

    def _build_holdings_map(
        self, transactions: list[dict], latest_prices: dict[str, float]
    ) -> dict[str, dict]:
        sorted_txns = sorted(transactions, key=lambda t: t["traded_at"])
        holdings: dict[str, dict] = {}

        for txn in sorted_txns:
            sym = txn["symbol"]
            h = holdings.get(sym, {
                "symbol": sym,
                "quantity": 0.0,
                "avg_price": 0.0,
                "current_price": latest_prices.get(sym, txn["price"]),
                "current_value": 0.0,
                "realized_pnl": 0.0,
                "unrealized_pnl": 0.0,
                "total_cost": 0.0,
            })

            if txn["side"] == "BUY":
                h["quantity"] += txn["quantity"]
                h["total_cost"] += txn["quantity"] * txn["price"]
            else:
                avg_cost = h["total_cost"] / h["quantity"] if h["quantity"] > 0 else 0
                h["realized_pnl"] += txn["quantity"] * (txn["price"] - avg_cost)
                h["quantity"] = max(0, h["quantity"] - txn["quantity"])
                h["total_cost"] = max(0, h["total_cost"] - avg_cost * txn["quantity"])

            h["avg_price"] = h["total_cost"] / h["quantity"] if h["quantity"] > 0 else 0
            h["current_price"] = latest_prices.get(sym, h["current_price"])
            h["current_value"] = h["quantity"] * h["current_price"]
            h["unrealized_pnl"] = h["quantity"] * (h["current_price"] - h["avg_price"])
            holdings[sym] = h

        return holdings

File: Server/src/portfolio/portfolioService.ts
import crypto from 'node:crypto'
import type { Pool } from 'pg'

type TransactionRow = {
  id: string
  symbol: string
  side: 'BUY' | 'SELL'
  quantity: string
  price: string
  traded_at: string
  metadata: Record<string, unknown>
}

export class PortfolioService {
  constructor(private readonly pool: Pool) {}

  async listTransactions(userId: string) {
    const result = await this.pool.query<TransactionRow>(
      `
        select id, symbol, side, quantity::text, price::text, traded_at::text, metadata
        from portfolio_transactions
        where user_id = $1
        order by traded_at desc, created_at desc
      `,
      [userId],
    )

    return result.rows.map((row) => ({
      id: row.id,
      symbol: row.symbol,
      side: row.side,
      quantity: Number(row.quantity),
      price: Number(row.price),
      tradedAt: row.traded_at,
      metadata: row.metadata ?? {},
    }))
  }

  async createTransaction(userId: string, input: { symbol: string; side: 'BUY' | 'SELL'; quantity: number; price: number; tradedAt?: string }) {
    const id = crypto.randomUUID()
    await this.pool.query(
      `
        insert into portfolio_transactions (id, user_id, symbol, side, quantity, price, traded_at)
        values ($1, $2, $3, $4, $5, $6, $7)
      `,
      [id, userId, input.symbol, input.side, input.quantity, input.price, input.tradedAt ?? new Date().toISOString()],
    )
    return id
  }

  async updateTransaction(userId: string, id: string, input: { quantity: number; price: number; tradedAt: string }) {
    await this.pool.query(
      `
        update portfolio_transactions
        set quantity = $3, price = $4, traded_at = $5
        where id = $1 and user_id = $2
      `,
      [id, userId, input.quantity, input.price, input.tradedAt],
    )
  }

  async deleteTransaction(userId: string, id: string) {
    await this.pool.query('delete from portfolio_transactions where id = $1 and user_id = $2', [id, userId])
  }

  async getSummary(userId: string, latestPrices: Record<string, number>) {
    const transactions = await this.listTransactions(userId)
    const holdingsMap = this.buildHoldingsMap(transactions, latestPrices)
    const holdings = Object.values(holdingsMap).filter((row) => row.quantity > 0)

    const invested = holdings.reduce((sum, row) => sum + row.avgPrice * row.quantity, 0)
    const current = holdings.reduce((sum, row) => sum + row.currentValue, 0)
    const realized = Object.values(holdingsMap).reduce((sum, row) => sum + row.realizedPnl, 0)
    const unrealized = holdings.reduce((sum, row) => sum + row.unrealizedPnl, 0)

    return {
      cards: [
        { id: 'invested', label: 'Total Invested', value: invested, change: null },
        { id: 'current', label: 'Current Value', value: current, change: invested ? (current / invested - 1) * 100 : 0 },
        { id: 'total_pl', label: 'Total P&L', value: realized + unrealized, change: invested ? ((realized + unrealized) / invested) * 100 : 0 },
        { id: 'unrealized', label: 'Unrealized', value: unrealized, change: invested ? (unrealized / invested) * 100 : 0 },
        { id: 'realized', label: 'Realized', value: realized, change: null },
      ],
      totals: {
        invested,
        current,
        realized,
        unrealized,
        totalPnl: realized + unrealized,
      },
    }
  }

  async getHoldings(userId: string, latestPrices: Record<string, number>) {
    const transactions = await this.listTransactions(userId)
    const holdingsMap = this.buildHoldingsMap(transactions, latestPrices)
    const rows = Object.values(holdingsMap)
      .filter((row) => row.quantity > 0)
      .map((row) => ({
        symbol: row.symbol,
        quantity: row.quantity,
        avgPrice: row.avgPrice,
        currentPrice: row.currentPrice,
        currentValue: row.currentValue,
        pnl: row.unrealizedPnl,
        pnlPct: row.avgPrice ? ((row.currentPrice - row.avgPrice) / row.avgPrice) * 100 : 0,
        realizedPnl: row.realizedPnl,
        allocationPct: 0,
      }))

    const totalCurrent = rows.reduce((sum, row) => sum + row.currentValue, 0)
    return rows.map((row) => ({
      ...row,
      allocationPct: totalCurrent ? (row.currentValue / totalCurrent) * 100 : 0,
    }))
  }

  private buildHoldingsMap(transactions: Awaited<ReturnType<PortfolioService['listTransactions']>>, latestPrices: Record<string, number>) {
    const ascending = [...transactions].sort((left, right) => new Date(left.tradedAt).getTime() - new Date(right.tradedAt).getTime())
    const holdings: Record<
      string,
      {
        symbol: string
        quantity: number
        avgPrice: number
        currentPrice: number
        currentValue: number
        realizedPnl: number
        unrealizedPnl: number
        totalCost: number
      }
    > = {}

    for (const txn of ascending) {
      const current = holdings[txn.symbol] ?? {
        symbol: txn.symbol,
        quantity: 0,
        avgPrice: 0,
        currentPrice: latestPrices[txn.symbol] ?? txn.price,
        currentValue: 0,
        realizedPnl: 0,
        unrealizedPnl: 0,
        totalCost: 0,
      }

      if (txn.side === 'BUY') {
        current.quantity += txn.quantity
        current.totalCost += txn.quantity * txn.price
      } else {
        const avgCost = current.quantity > 0 ? current.totalCost / current.quantity : 0
        current.realizedPnl += txn.quantity * (txn.price - avgCost)
        current.quantity = Math.max(0, current.quantity - txn.quantity)
        current.totalCost = Math.max(0, current.totalCost - avgCost * txn.quantity)
      }

      current.avgPrice = current.quantity > 0 ? current.totalCost / current.quantity : 0
      current.currentPrice = latestPrices[txn.symbol] ?? current.currentPrice
      current.currentValue = current.quantity * current.currentPrice
      current.unrealizedPnl = current.quantity * (current.currentPrice - current.avgPrice)
      holdings[txn.symbol] = current
    }

    return holdings
  }
}
